### PR TITLE
Configs

### LINKED ISSUE #8: Move Customizable Constants to Config File
Currently, the prefix and other details are only defined in the `main.js`. This makes it inaccessible to any plugins easily and makes them hard to change.

- [ ] Move/add the following constants from `main.js` to a `.json` config file. Convert to camelCase.
  - [ ] `PREFIX`
  - [ ] Activity
    - [ ] Text
    - [ ] Type
  - [ ] Default Color
  - [ ] Enabled Plugins (an array of strings representing slugs)
  - [ ] Winston Discord log level (log files should remain hard-coded where they are, and console log level should remain as an env variable, as that can feasibly change run-to-run)
  - [ ] Logging Discord Webhook activation (whether or not logs will be sent to the server)
- [ ] Create a template config to commit (like `.env.template`), then add to `.gitignore` the name of the final config file so it's never committed.
- [ ] Add functions to `utils.js` to retrieve and set these constants. Cache in a variable, but persist to disk.
- [ ] Update `main.js` and `utils.js` to use these functions.
- [ ] Comment in the Issue any changes that need to be made to the blueprint plugins.
- [ ] Update the `Makefile` and `package.json` to allow easily starting the testing bot. See the [Spike-2](https://github.com/spike-2/Spike-2) versions of these files as a reference, but be sure to change the prefix in the config file instead of `main.js`

### PR DESCRIPTION
**Pull request by jwMaxwell**
This PR is in reference to Issue #8 

(Pasting from below)
### Config generation
- When you start the bot, it should generate a config file in the project directory
- If no config is present, the bot should also log that no config was found
### Config use
- Change the values of `activity.type` and `activity.description`. (Valid activity types are WATCHING, PLAYING, LISTENING, STREAMING, and all values should be case-insensitive)
- Change the value of prefix and try sending a message that the bot will respond to.
- Change the values of discordLogging.active and discordLogging.level. Ensure that the logging is affected accordingly. (Valid logging levels are error, warn, info, verbose, debug, silly)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -12,6 +12,9 @@ yarn-error.log*
 lerna-debug.log*
 .pnpm-debug.log*
 
+# Configuration file
+config.json*
```
Comment (brandoningli) [inline .gitignore]: This should just match `config.json`. The current selector will also grab things like a future `config.json.template` we may add.

File: src/configManager.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    );
+  });
+
+  return defaultConfig;
+}
+
+function loadConfig() {
+  try {
+    return JSON.parse(
+      fs.readFileSync("../config.json", { encoding: "utf8", flag: "r" })
```
Comment (brandoningli) [inline src/configManager.js:22]: Either move the `.env` file to the parent directory or keep `config.json` in `/src`.

File: src/configManager.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      `config.json not found. Generating new configuration file...`
+    );
+  });
+
+  return defaultConfig;
+}
+
+function loadConfig() {
+  try {
+    return JSON.parse(
```
Comment (brandoningli) [inline src/configManager.js]: Is there a reason why this isn't just initializing `config` directly instead of returning it? Especially since `config` is exported.

File: src/main.js [hunk 4]
```diff
@@ -100,6 +85,28 @@ utils.logger = winston.createLogger({
   ],
 });
 
+// config must be loaded after the logger is initialized
+config = loadConfig();
+
+if (config.discordLogging.active.toUpperCase() === "TRUE") {
```
Comment (brandoningli) [inline src/main.js]: All references to config values need a null coalesce (`??`) with a default value that makes sense. This may be able to be abstracted away with a function?

File: src/configManager.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+const fs = require("fs");
+
+let config;
+
+const defaultConfig = {
+  prefix: "$",
+  activity: { type: "playing", description: "$help" },
+  color: "purple",
+  plugins: {},
+  discordLogging: { active: "true", level: "warn" },
```
Comment (brandoningli) [inline src/configManager.js]: `active` should be a bool, not a string.

**Review by brandoningli — CHANGES_REQUESTED** [review]
Additional notes:

- Please add testing instructions
- Issue description calls for a default template file to be committed. Please do so.
- `utils.js` was not updated to use the new config
- Double check the template plugins in the wiki, and indicate if anything (or nothing) needs to change to accommodate these changes.
- Update the `Makefile` and `package.json` to allow for easily running testing
- Have not yet run the code to test functionality; just a code review performed

**Comment by jwMaxwell** [comment]
## Testing
### Config generation
- When you start the bot, it should generate a config file in the project directory
- If no config is present, the bot should also log that no config was found
### Config use
- Change the values of `activity.type` and `activity.description`. (Valid activity types are WATCHING, PLAYING, LISTENING, STREAMING, and all values should be case-insensitive)
- Change the value of prefix and try sending a message that the bot will respond to.
- Change the values of discordLogging.active and discordLogging.level. Ensure that the logging is affected accordingly. (Valid logging levels are error, warn, info, verbose, debug, silly)

File: src/main.js [hunk 6]
```diff
@@ -114,8 +121,11 @@ const bot = new Client({
 
 bot.on("ready", async () => {
   // when loaded (ready event)
-  bot.user.setActivity(`${PREFIX}help | ${PREFIX}info`, { type: "PLAYING" });
+  bot.user.setActivity(config.activity.description, {
+    type: config.activity.type.toUpperCase(),
+  });
   utils.logger.log("debug", `${bot.user.username} is ready...`);
+  console.log(`${bot.user.username} is ready...`);
```
Comment (brandoningli) [inline src/main.js]: Per the style guide, console.log should not be in production code. It's already being logged at the debug level. If you think it should be logged at a more severe level, update that instead.

**Review by brandoningli — CHANGES_REQUESTED** [review]
Other than adding testing instructions, none of the previous comments have been addressed.

File: src/configManager.js (hunk 3 above)
Comment (jwMaxwell) [reply src/configManager.js]: Yes. Initializing `config` directly doesn't work due to scoping issues (I believe). When the bot is started, you will receive this error:
```
[2022-04-24 15:49:08] [error] uncaughtException: Cannot read properties of undefined (reading 'discordLogging')
TypeError: Cannot read properties of undefined (reading 'discordLogging')
    at Object.<anonymous> (/home/joshua/Workspace/Modibo/src/main.js:91:12)
    at Module._compile (node:internal/modules/cjs/loader:1099:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:975:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at node:internal/main/run_main_module:17:47
```

Comment (brandoningli) [reply src/configManager.js]: This may be due to the destructuring that happens. Does importing the module as a single `const` allow you to then both call those functions and access the `config` property?

```js
const configManager = require("./configManager");
...
configManager.loadConfig();
...

if (configManager.config.discordLogging.active...)
```

Comment (jwMaxwell) [reply src/configManager.js]: Same error

Comment (brandoningli) [reply src/configManager.js:28]: Referring to the `config` variable within the `loadConfig()` function as `module.config` might do it?

File: src/configManager.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    utils.logger.log(
+      "warn",
+      `config.json not found. Generating new configuration file...`
+    );
+  });
+
+  return defaultConfig;
+}
+
+// function loadConfig() {
```
Comment (brandoningli) [inline src/configManager.js]: Please remove the commented out code

File: src/configManager.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  activity: { type: "playing", description: "$help" },
+  color: "purple",
+  plugins: {},
+  discordLogging: { active: true, level: "warn" },
+};
+
+function initConfig() {
+  fs.writeFile("../config.json", JSON.stringify(defaultConfig), function (err) {
+    if (err) throw "Error auto-generating configuration file.";
+    utils.logger.log(
```
Comment (brandoningli) [inline src/configManager.js]: This is not run when creating a default config. Move outside the error callback for the `writeFile` call so it always runs when `initConfig` runs

File: src/main.js [hunk 9]
```diff
@@ -100,6 +85,28 @@ utils.logger = winston.createLogger({
   ],
 });
 
+// config must be loaded after the logger is initialized
+configManager.loadConfig();
+
+if (configManager.config.discordLogging.active ?? true) {
```
Comment (brandoningli) [inline src/main.js]: When config is completely empty, error. When the `discordLogging` key is defined, all is good.
```
[2022-04-24 18:28:08] [error] uncaughtException: Cannot read properties of undefined (reading 'active')
TypeError: Cannot read properties of undefined (reading 'active')
    at Object.<anonymous> (/home/brandon/Documents/git/devcodeabode/MoDiBo/src/main.js:91:41)
    at Module._compile (node:internal/modules/cjs/loader:1103:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1155:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at node:internal/main/run_main_module:17:47
    ```

File: src/main.js [hunk 10]
```diff
@@ -114,8 +121,11 @@ const bot = new Client({
 
 bot.on("ready", async () => {
   // when loaded (ready event)
-  bot.user.setActivity(`${PREFIX}help | ${PREFIX}info`, { type: "PLAYING" });
+  bot.user.setActivity(configManager.config.activity.description ?? "nil", {
+    type: configManager.config.activity.type.toUpperCase() ?? "PLAYING",
```
Comment (brandoningli) [inline src/main.js]: Need to move the `??` so it's in parens with `configManager.config.activity.type`. When activity key is empty but present:
```
[2022-04-24 18:29:13] [error] uncaughtException: Cannot read properties of undefined (reading 'toUpperCase')
TypeError: Cannot read properties of undefined (reading 'toUpperCase')
    at Client.<anonymous> (/home/brandon/Documents/git/devcodeabode/MoDiBo/src/main.js:125:46)
    at Client.emit (node:events:520:28)
    at WebSocketManager.triggerClientReady (/home/brandon/Documents/git/devcodeabode/MoDiBo/node_modules/discord.js/src/client/websocket/WebSocketManager.js:384:17)
    at WebSocketManager.checkShardsReady (/home/brandon/Documents/git/devcodeabode/MoDiBo/node_modules/discord.js/src/client/websocket/WebSocketManager.js:367:10)
    at WebSocketShard.<anonymous> (/home/brandon/Documents/git/devcodeabode/MoDiBo/node_modules/discord.js/src/client/websocket/WebSocketManager.js:189:14)
    at WebSocketShard.emit (node:events:520:28)
    at WebSocketShard.checkReady (/home/brandon/Documents/git/devcodeabode/MoDiBo/node_modules/discord.js/src/client/websocket/WebSocketShard.js:475:12)
    at WebSocketShard.onPacket (/home/brandon/Documents/git/devcodeabode/MoDiBo/node_modules/discord.js/src/client/websocket/WebSocketShard.js:447:16)
    at WebSocketShard.onMessage (/home/brandon/Documents/git/devcodeabode/MoDiBo/node_modules/discord.js/src/client/websocket/WebSocketShard.js:301:10)
    at WebSocket.onMessage (/home/brandon/Documents/git/devcodeabode/MoDiBo/node_modules/ws/lib/event-target.js:199:18)
    ```

File: src/main.js [hunk 11]
```diff
@@ -131,9 +141,13 @@ bot.on("messageCreate", async (message) => {
   }
 
   // if it is a command
-  if (message.content.charAt(0) === PREFIX) {
+  if (message.content.charAt(0) === configManager.config.prefix ?? "$") {
```
Comment (brandoningli) [inline src/main.js]: With no prefix key or an empty string, commands don't work at all. `$err` et. al. don't even produce console output. Perhaps use `||` in these situations to account for empty strings?

File: src/main.js [hunk 12]
```diff
@@ -114,8 +121,11 @@ const bot = new Client({
 
 bot.on("ready", async () => {
   // when loaded (ready event)
-  bot.user.setActivity(`${PREFIX}help | ${PREFIX}info`, { type: "PLAYING" });
+  bot.user.setActivity(configManager.config.activity.description ?? "nil", {
+    type: configManager.config.activity.type.toUpperCase() ?? "PLAYING",
+  });
   utils.logger.log("debug", `${bot.user.username} is ready...`);
+  console.info(`${bot.user.username} is ready...`);
```
Comment (brandoningli) [inline src/main.js:159]: Making an exception to the "No `console.log`" style guide rule for this instance. Regardless of the console log level, there should be some feedback in the console that the bot has started.

**Review by brandoningli — CHANGES_REQUESTED** [review]
- Issue description calls for a default template file to be committed. Please do so, or add to the README describing how to generate a config.
- utils.js was not updated to use the new config
- Update the Makefile and package.json to allow for easily running testing
- Move `.env.template` to the root directory since that's where it'll be read from now. `git mv src/.env.template .` so git will register it as a rename, not a delete/add.

**Review by brandoningli — DISMISSED** [review]
I made these changes, so I approve. Would like another approval before this is merged.

**Review by brandoningli — DISMISSED** [review]

**Review by brandoningli — APPROVED** [review]

**Review by brandoningli — APPROVED** [review]
